server/adaptors/integrations/repository/integration_reader.ts

```diff
@@ -0,0 +1,593 @@
+  private async fetchDataOrReadFile(
+    item: { data?: string },
+    fileParams: { filename: string; type?: IntegrationPart },
+    format: 'binary'
+  ): Promise<Result<Buffer>>;
+  private async fetchDataOrReadFile(
+    item: { data?: string },
+    fileParams: { filename: string; type?: IntegrationPart },
+    format: 'json' | 'binary'
+  ): Promise<Result<object | object[] | Buffer>> {
+    if (this.reader.isConfigLocalized) {
+      if (!item.data) {
+        return {
+          ok: false,
+          error: new Error(
+            'The config for the provided reader is localized, but no data field is present. ' +
+              JSON.stringify(item)
+          ),
+        };
+      }
+      try {
+        if (format === 'json') {
+          return { ok: true, value: JSON.parse(item.data) };
+        } else {
+          return { ok: true, value: Buffer.from(item.data, 'base64') };
+        }
+      } catch (error) {
+        return { ok: false, error };
+      }
+    }
+
+    if (format === 'json') {
+      return this.reader.readFile(fileParams.filename, fileParams.type);
+    } else {
+      return this.reader.readFileRaw(fileParams.filename, fileParams.type);
+    }
+  }
+
+  /**
+   * Like getConfig(), but thoroughly checks all nested integration dependencies for validity.
+   *
+   * @returns a Result indicating whether the integration is valid.
+   */
+  async deepCheck(): Promise<Result<IntegrationConfig>> {
+    const configResult = await this.getConfig();
+    if (!configResult.ok) {
+      return configResult;
+    }
+
+    // Some other checks are included in getConfig validation.
+    const assets = await this.getAssets();
+    if (!assets.ok || Object.keys(assets.value).length === 0) {
+      return { ok: false, error: new Error('An integration must have at least one asset') };
+    }
+
+    return configResult;
+  }
+
+  /**
+   * Get the latest version of the integration available.
+   * This method relies on the fact that integration configs have their versions in their name.
+   * Any files that don't match the config naming convention will be ignored.
+   *
+   * @returns A string with the latest version, or null if no versions are available.
+   */
+  async getLatestVersion(): Promise<string | null> {
+    const versions = await this.reader.findIntegrationVersions();
+    if (!versions.ok) {
+      return null;
+    }
+    if (versions.value.length === 0) {
+      return null;
+    }
+    // Sort descending
+    versions.value.sort(semver.rcompare);
+    return versions.value[0];
+  }
+
+  // Get config without pruning or validation.
+  private async getRawConfig(
+    version?: string
+  ): Promise<Result<IntegrationConfig | SerializedIntegration>> {
+    if ((await this.reader.getDirectoryType()) !== 'integration') {
+      return { ok: false, error: new Error(`${this.directory} is not a valid integration`) };
+    }
+
+    const maybeVersion: string | null = version ? version : await this.getLatestVersion();
```

Comment:
It's just validation. Integrations should always have at least one valid version, but there are configuration error scenarios where this might not be true (where we return an error on the next line).